idaes/apps/grid_integration/tracker.py

```diff
@@ -0,0 +1,1747 @@
+        m.slack_var_power.fix(0)
+
+        return
+
+    @staticmethod
+    def _add_min_slack_objective(m):
+
+        def min_slack_fun(m):
+            return sum(m.slack_var_power[j,h] + m.pbc_slack_positive[j,h] + \
+                   m.pbc_slack_negative[j,h] for j in m.UNITS for h in m.HOUR)
+        m.min_slack = pyo.Objective(rule = min_slack_fun, sense = pyo.minimize)
+        m.min_slack.deactivate()
+
+        return
+
+    @staticmethod
+    def _add_slack_constraints(m):
+
+        plan_horizon = len(m.HOUR)
+
+        ## add constraints for power dispatch
+        def abs_con_fun1(m,j,h):
+            return -m.slack_var_power[j,h]<=m.P_total[j,h] - m.power_dispatch[j,h]
+        m.power_slack_con1 = pyo.Constraint(m.UNITS,m.HOUR,rule = abs_con_fun1)
+        m.power_slack_con1.deactivate()
+
+        def abs_con_fun2(m,j,h):
+            return m.slack_var_power[j,h]>=m.P_total[j,h] - m.power_dispatch[j,h]
+        m.power_slack_con2 = pyo.Constraint(m.UNITS,m.HOUR,rule = abs_con_fun2)
+        m.power_slack_con2.deactivate()
+
+        ## constraints for pbc slacks
+        def pbc_slack_fun(m,j,h):
+            return m.pbc_slack[j,h] == m.pbc_slack_positive[j,h] - m.pbc_slack_negative[j,h]
+        m.pbc_slack_con = pyo.Constraint(m.UNITS,m.HOUR,rule = pbc_slack_fun)
+        m.pbc_slack_con.deactivate()
+
+        def SOC_endtime_slack_fun(m,j,h):
+            if h == plan_horizon - 1:
+                return  m.S_SOC[j,h] == m.pre_SOC[j] + m.pbc_slack[j,h]
+            else:
+                return pyo.Constraint.Skip
+        m.SOC_endtime_slack_con = pyo.Constraint(m.UNITS,m.HOUR,rule = SOC_endtime_slack_fun)
+        m.SOC_endtime_slack_con.deactivate()
+
+        return
+
+    @staticmethod
+    def _add_slack_bounding_constraints(m):
+
+        # add minimum pbc slack as a param
+        m.pbc_slack_bound = pyo.Param(m.UNITS, initialize = 0, mutable = True)
+
+        def bound_pbc_slack_fun(m):
+            for j in m.UNITS:
+                yield sum(m.pbc_slack[j,h] for h in m.HOUR) <= 1.01 * m.pbc_slack_bound[j]
+                yield sum(m.pbc_slack[j,h] for h in m.HOUR) >= -1.01 * m.pbc_slack_bound[j]
+        m.bound_pbc_slack_con = pyo.ConstraintList(rule = bound_pbc_slack_fun)
+        m.bound_pbc_slack_con.deactivate()
+
+        # add minimum power output slack as a param
+        m.power_slack_bound = pyo.Param(m.UNITS, m.HOUR, initialize = 0, mutable = True)
+
+        def bound_power_slack_fun(m):
+            for j in m.UNITS:
+                for h in m.HOUR:
+                    yield m.slack_var_power[j,h]  <= 1.01 * m.power_slack_bound[j,h]
+                    yield m.slack_var_power[j,h]  >= -1.01 * m.power_slack_bound[j,h]
+        m.bound_power_slack_con = pyo.ConstraintList(rule = bound_power_slack_fun)
+        m.bound_power_slack_con.deactivate()
+
+        return
+
+    @staticmethod
+    def _build_model(model_data,plan_horizon,segment_number=4):
+
```

Comment:
How can we reduce the amount of duplicate code for this example when defining the bidder model and tracker model objects?